Execute subqueries only once

mrzzzrm opened the pull request:
Previously the LQPTranslator would always translate an LQPSelectExpression to a PQPSelectExpression. Now it first checks whether the LQPSelectExpression was already available in an output column of the input node and creates a PQPColumnExpression pointing to this column if it is.

Performance with low scale factor, so TPC-H 20 could finish a few times

```
s=1.0
+-----------+-------------------+------+-------------------+------+--------+---------------------------------+
| Benchmark | prev. iter/s      | runs | new iter/s        | runs | change | p-value (significant if <0.001) |
+-----------+-------------------+------+-------------------+------+--------+---------------------------------+
| TPC-H 1   | 0.106872729957    | 20   | 0.107606701553    | 20   | +1%    |                          0.0000 |
| TPC-H 2   | 0.530021131039    | 20   | 0.544726490974    | 20   | +3%    |     (run time too short) 0.0000 |
| TPC-H 3   | 0.865605711937    | 20   | 0.866196155548    | 20   | +0%    |     (run time too short) 0.6583 |
| TPC-H 4   | 0.00102076807525  | 3    | 0.00105136760976  | 3    | +3%    |        (not enough runs) 0.0000 |
| TPC-H 5   | 0.639450192451    | 20   | 0.635892450809    | 20   | -1%    |     (run time too short) 0.0814 |
| TPC-H 6   | 1.35202097893     | 20   | 1.30782902241     | 20   | -3%    |     (run time too short) 0.0000 |
| TPC-H 7   | 0.137286990881    | 20   | 0.136643409729    | 20   | -0%    |                          0.2520 |
| TPC-H 8   | 0.0135615486652   | 20   | 0.0140657797456   | 20   | +4%    |                          0.0014 |
| TPC-H 9   | 0.0725982189178   | 20   | 0.0744021236897   | 20   | +2%    |                          0.0044 |
| TPC-H 10  | 0.487696647644    | 20   | 0.515312254429    | 20   | +6%    |     (run time too short) 0.0000 |
| TPC-H 11  | 1.84367930889     | 20   | 2.05625796318     | 20   | +12%   |     (run time too short) 0.0000 |
| TPC-H 12  | 0.751853466034    | 20   | 0.834767103195    | 20   | +11%   |     (run time too short) 0.0000 |
| TPC-H 13  | 0.752767562866    | 20   | 0.820633947849    | 20   | +9%    |     (run time too short) 0.0000 |
| TPC-H 14  | 1.50283622742     | 20   | 1.64700460434     | 20   | +10%   |     (run time too short) 0.0000 |
| TPC-H 15  | 0.425549417734    | 20   | 0.449809104204    | 20   | +6%    |     (run time too short) 0.0000 |
| TPC-H 16  | 0.392209976912    | 20   | 0.411731928587    | 20   | +5%    |     (run time too short) 0.0010 |
| TPC-H 17  | 0.00904178805649  | 19   | 0.00982083939016  | 20   | +9%    |                          0.0000 |
| TPC-H 18  | 0.00157690455671  | 4    | 0.00164479378145  | 4    | +4%    |        (not enough runs) 0.3601 |
| TPC-H 19  | 0.0638135001063   | 20   | 0.064276188612    | 20   | +1%    |                          0.0000 |
| TPC-H 20  | 4.22515149694e-05 | 1    | 7.40073737688e-05 | 1    | +75%   |           (not enough runs) nan |
| TPC-H 21  | 0.000711627944838 | 2    | 0.000739644165151 | 2    | +4%    |        (not enough runs) 0.0001 |
| TPC-H 22  | 0.00161762768403  | 4    | 0.00166534201708  | 4    | +3%    |        (not enough runs) 0.0000 |
| average   |                   |      |                   |      | +7%    |                                 |
+-----------+-------------------+------+-------------------+------+--------+---------------------------------+

```

mrks commented:
changed "fixes" to "addresses", because in any case, the visualization part is still unsolved.

mrks commented:
This seems to have fixed it for
`SELECT ps_suppkey FROM partsupp WHERE ps_partkey IN (SELECT 2) AND ps_availqty > (SELECT 1);`
but when the subselect is reused, it is now broken:
```
(debug)> visualize  SELECT ps_suppkey FROM partsupp WHERE ps_partkey IN (SELECT 1) AND ps_availqty > (SELECT 1);
Unsupported ExpressionType used in InExpression
```

Also, any idea how to test this?

mrzzzrm commented:
* Fixed `SELECT ps_suppkey FROM partsupp WHERE ps_partkey IN (SELECT 1) AND ps_availqty > (SELECT 1);` by making the `InExpression` compatible with all kinds of operands.

* Added tests for column reuse

mrks commented on src/lib/expression/evaluation/expression_evaluator.cpp:
```diff
@@ -336,7 +336,13 @@ ExpressionEvaluator::_evaluate_in_expression<ExpressionEvaluator::Bool>(const In
+     * To accomplish this, we simply rewrite the expression to `<expression> IN LIST(<anything>)`
+     */
+
+    return _evaluate_in_expression<ExpressionEvaluator::Bool>(*in_(in_expression.value(), list_(in_expression.set())));
```
I don't understand this yet. Why did this become necessary now? What does this mean for this branch:

https://github.com/hyrise/hyrise/blob/c83d85b533688de22defbc56b56db1cbb69ac65a/src/lib/expression/evaluation/expression_evaluator.cpp#L284

Also, does this mean that you now build that long `OR` list instead of going through the result?

mrks replied:
To clarify: I understand that PQPSelects are now rewritten to PQPColumns, correct? But wouldn't we still want to handle them the same way?

mrzzzrm commented:
Alright, to clear up the question why the result of `(SELECT 1)` ended up in a column for the query `SELECT ps_suppkey FROM partsupp WHERE ps_partkey IN (SELECT 1) AND ps_availqty > (SELECT 1);`

The SQLTranslator plans the predicate `ps_availqty > (SELECT 1)` first. Since one of the arguments (`(SELECT 1)`) is not a column or a literal, it gets computed into a column. (This place in the plan is obviously where an Assert would be triggered if `(SELECT 1)` would have returned more than one row or column during execution.)
Then, a column-vs-column scan is performed for `ps_availqty > (SELECT 1)`.

Since the TableScan cannot deal with `IN`, `ps_partkey IN (SELECT 1)` is being computed by a Projection into a bool-column. When this Projection is translated in the LQPTranslator, `(SELECT 1)` is already available in a column, thus  `ps_partkey IN (SELECT 1)` is rewritten to  `ps_partkey IN <some column containing SELECT 1>`

mrzzzrm commented:
if CI thinks this is okay, I think we can merge
